File: apps/web/src/api/faculty.ts
import { apiClient } from "./client";

export interface FacultyDashboard {
  has_scope: boolean;
  first_name: string;
  department_name: string | null;
  semester_label: string;
  session_label: string;
  day_summary: string;
  health_score: number;
  health_label: string;
  health_factors: Array<{ label: string; value: number }>;
  health_narrative: string;
  health_sources: string[];
  daily_brief: { text: string; bullets: string[] };
  brief_actions: Array<{ icon: string; text: string; link_label: string }>;
  kpis: Array<{ id: string; label: string; value: string; sub?: string | null }>;
  classes_today: Array<{
    id: string;
    start_time: string;
    title: string;
    section: string | null;
    room: string | null;
    student_count: number | null;
    status: string;
    status_note: string | null;
    session_type: string;
  }>;
  coaching_items: Array<{ title: string; why: string; cta: string; coach_key: string }>;
  attendance_tasks: {
    pending: number;
    completed: number;
    overdue: number;
    filled_count: number;
    total_count: number;
    registers: Array<{ section: string; course_code: string; start_time: string; status: string }>;
  } | null;
  class_performance: {
    avg_internal_pct: number;
    delta_pts: number | null;
    internals_in_days: number | null;
    subjects: Array<{ section: string; course_code: string; avg_pct: number; tone: string }>;
    insight: string;
  } | null;
  assignment_reviews: Array<{
    id: string;
    title: string;
    priority: string;
    due_label: string;
    section: string;
    submission_count: number;
    graded_count: number;
  }>;
  pending_approvals: Array<{ id: string; kind: string; title: string; due_label: string }>;
  flagged_students: Array<{
    student_id: string;
    name: string;
    section: string;
    why: string;
    based_on: string;
    confidence_pct: number;
    confidence_label: string;
    tier: string;
  }>;
  nav_badges: { reviews: number; students: number; approvals: number };
  at_risk_students: Array<{
    student_id: string;
    name: string;
    roll_no: string;
    tier: string;
    top_finding: string | null;
  }>;
  attendance_watch: Array<{
    student_id: string;
    name: string;
    roll_no: string;
    present_rate_pct: number | null;
    note: string;
  }>;
  course_progress: Array<{
    course_id: string;
    course_code: string;
    course_name: string;
    coverage_pct: number;
    planned_sessions: number;
    delivered_sessions: number;
    slippage: boolean;
  }>;
  recent_activity: Array<{ id: string; kind: string; text: string; at: string }>;
  import_freshness: { is_stale: boolean; message: string | null } | null;
}

export interface FacultyArtifact {
  id: string;
  artifact_type: string;
  title: string;
  status: string;
  content_json: Record<string, unknown>;
  source_job_id: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface GenerationJob {
  id: string;
  feature: string;
  params_json: Record<string, unknown>;
  status: string;
  result_artifact_id: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface CoursePlan {
  id: string;
  course_id: string;
  course_code: string;
  course_name: string;
  syllabus_units: unknown[];
  planned_sessions: number;
  delivered_sessions: number;
}

export interface OfficeHourSlot {
  id: string;
  slot_date: string;
  start_time: string;
  end_time: string;
  location: string | null;
  status: string;
  booking_note: string | null;
}

export async function getFacultyDashboard(): Promise<FacultyDashboard> {
  const { data } = await apiClient.get<FacultyDashboard>("/faculty/dashboard");
  return data;
}

export async function listFacultyArtifacts(): Promise<FacultyArtifact[]> {
  const { data } = await apiClient.get<FacultyArtifact[]>("/faculty/artifacts");
  return data;
}

export async function getFacultyArtifact(id: string): Promise<FacultyArtifact> {
  const { data } = await apiClient.get<FacultyArtifact>(`/faculty/artifacts/${id}`);
  return data;
}

export async function updateFacultyArtifact(
  id: string,
  payload: { title?: string; status?: string; content_json?: Record<string, unknown> },
): Promise<FacultyArtifact> {
  const { data } = await apiClient.patch<FacultyArtifact>(`/faculty/artifacts/${id}`, payload);
  return data;
}

export async function generateFacultyContent(
  feature: string,
  params: Record<string, unknown> = {},
): Promise<GenerationJob> {
  const { data } = await apiClient.post<GenerationJob>("/faculty/generate", { feature, params });
  return data;
}

export async function getGenerationJob(id: string): Promise<GenerationJob> {
  const { data } = await apiClient.get<GenerationJob>(`/faculty/jobs/${id}`);
  return data;
}

export async function listCoursePlans(): Promise<CoursePlan[]> {
  const { data } = await apiClient.get<CoursePlan[]>("/faculty/course-plans");
  return data;
}

export async function listOfficeHours(): Promise<OfficeHourSlot[]> {
  const { data } = await apiClient.get<OfficeHourSlot[]>("/faculty/office-hours");
  return data;
}

export async function createOfficeHour(payload: {
  slot_date: string;
  start_time: string;
  end_time: string;
  location?: string;
}): Promise<OfficeHourSlot> {
  const { data } = await apiClient.post<OfficeHourSlot>("/faculty/office-hours", payload);
  return data;
}

File: apps/api/app/schemas/faculty.py
"""Faculty workspace API schemas."""

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


class DailyBriefPayload(BaseModel):
    text: str
    bullets: list[str]


class FacultyKpiItem(BaseModel):
    id: str
    label: str
    value: str
    sub: str | None = None


class AtRiskStudentItem(BaseModel):
    student_id: UUID
    name: str
    roll_no: str
    tier: str
    top_finding: str | None = None


class AttendanceWatchItem(BaseModel):
    student_id: UUID
    name: str
    roll_no: str
    present_rate_pct: float | None = None
    note: str


class CourseProgressItem(BaseModel):
    course_id: UUID
    course_code: str
    course_name: str
    coverage_pct: float
    planned_sessions: int
    delivered_sessions: int
    slippage: bool


class RecentActivityItem(BaseModel):
    id: str
    kind: str
    text: str
    at: datetime


class ImportFreshness(BaseModel):
    is_stale: bool
    message: str | None = None


class HealthFactor(BaseModel):
    label: str
    value: int


class CoachItem(BaseModel):
    title: str
    why: str
    cta: str
    coach_key: str


class BriefActionItem(BaseModel):
    icon: str
    text: str
    link_label: str


class FacultyClassSession(BaseModel):
    id: str
    start_time: str
    title: str
    section: str | None = None
    room: str | None = None
    student_count: int | None = None
    status: str
    status_note: str | None = None
    session_type: str


class AttendanceRegisterItem(BaseModel):
    section: str
    course_code: str
    start_time: str
    status: str


class AttendanceTasksPayload(BaseModel):
    pending: int
    completed: int
    overdue: int
    filled_count: int
    total_count: int
    registers: list[AttendanceRegisterItem]


class ClassPerformanceSubject(BaseModel):
    section: str
    course_code: str
    avg_pct: float
    tone: str


class ClassPerformancePayload(BaseModel):
    avg_internal_pct: float
    delta_pts: float | None = None
    internals_in_days: int | None = None
    subjects: list[ClassPerformanceSubject]
    insight: str


class AssignmentReviewItem(BaseModel):
    id: str
    title: str
    priority: str
    due_label: str
    section: str
    submission_count: int
    graded_count: int


class ApprovalItem(BaseModel):
    id: str
    kind: str
    title: str
    due_label: str


class NavBadges(BaseModel):
    reviews: int
    students: int
    approvals: int


class FlaggedStudentItem(BaseModel):
    student_id: UUID
    name: str
    section: str
    why: str
    based_on: str
    confidence_pct: int
    confidence_label: str
    tier: str


class FacultyDashboardResponse(BaseModel):
    has_scope: bool
    first_name: str = ""
    department_name: str | None = None
    semester_label: str = ""
    session_label: str = ""
    day_summary: str = ""
    health_score: int = 0
    health_label: str = ""
    health_factors: list[HealthFactor] = Field(default_factory=list)
    health_narrative: str = ""
    health_sources: list[str] = Field(default_factory=list)
    daily_brief: DailyBriefPayload
    brief_actions: list[BriefActionItem] = Field(default_factory=list)
    kpis: list[FacultyKpiItem]
    classes_today: list[FacultyClassSession] = Field(default_factory=list)
    coaching_items: list[CoachItem] = Field(default_factory=list)
    attendance_tasks: AttendanceTasksPayload | None = None
    class_performance: ClassPerformancePayload | None = None
    assignment_reviews: list[AssignmentReviewItem] = Field(default_factory=list)
    pending_approvals: list[ApprovalItem] = Field(default_factory=list)
    flagged_students: list[FlaggedStudentItem] = Field(default_factory=list)
    nav_badges: NavBadges = Field(default_factory=lambda: NavBadges(reviews=0, students=0, approvals=0))
    at_risk_students: list[AtRiskStudentItem]
    attendance_watch: list[AttendanceWatchItem]
    course_progress: list[CourseProgressItem]
    recent_activity: list[RecentActivityItem]
    import_freshness: ImportFreshness | None = None


class FacultyArtifactResponse(BaseModel):
    id: UUID
    artifact_type: str
    title: str
    status: str
    content_json: dict
    source_job_id: UUID | None
    version: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class FacultyArtifactCreate(BaseModel):
    artifact_type: str = "draft"
    title: str = "Untitled draft"
    content_json: dict = Field(default_factory=dict)


class FacultyArtifactUpdate(BaseModel):
    title: str | None = None
    status: str | None = None
    content_json: dict | None = None


class GenerationJobResponse(BaseModel):
    id: UUID
    feature: str
    params_json: dict
    status: str
    result_artifact_id: UUID | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class GenerateRequest(BaseModel):
    feature: str
    params: dict = Field(default_factory=dict)


class OfficeHourSlotResponse(BaseModel):
    id: UUID
    slot_date: date
    start_time: str
    end_time: str
    location: str | None
    status: str
    booking_note: str | None

    model_config = {"from_attributes": True}


class OfficeHourSlotCreate(BaseModel):
    slot_date: date
    start_time: str
    end_time: str
    location: str | None = None


class CoursePlanResponse(BaseModel):
    id: UUID
    course_id: UUID
    course_code: str
    course_name: str
    syllabus_units: list
    planned_sessions: int
    delivered_sessions: int

    model_config = {"from_attributes": True}


class InterventionCreateRequest(BaseModel):
    student_id: UUID
    type: str
    title: str
    notes: str | None = None
